search-service/src/main/kotlin/com/egm/stellio/search/csr/service/ContextSourceUtils.kt

```diff
@@ -26,9 +26,34 @@ import java.time.ZonedDateTime
 import kotlin.random.Random.Default.nextBoolean
 
 typealias CompactedEntityWithCSR = Pair<CompactedEntity, ContextSourceRegistration>
+typealias CompactedEntitiesWithCSR = Pair<List<CompactedEntity>, ContextSourceRegistration>
+
 typealias AttributeByDatasetId = Map<String?, CompactedAttributeInstance>
 object ContextSourceUtils {
 
+    fun mergeEntitiesLists(
+        localEntities: List<CompactedEntity>,
+        remoteEntitiesWithCSR: List<CompactedEntitiesWithCSR>
+    ): IorNel<NGSILDWarning, List<CompactedEntity>> {
+        val mergedEntityMap = localEntities.map { it.toMutableMap() }.associateBy { it[JSONLD_ID_TERM] }.toMutableMap()
+
+        val warnings = remoteEntitiesWithCSR.sortedBy { (_, csr) -> csr.isAuxiliary() }.mapNotNull { (entities, csr) ->
+            val test = either {
+                entities.forEach { entity ->
+                    val id = entity[JSONLD_ID_TERM]
+                    mergedEntityMap[id]
+                        ?.let { it.putAll(getMergeNewValues(it, entity, csr).bind()) }
+                        ?: run { mergedEntityMap[id] = entity.toMutableMap() }
+                }
+                null
+            }
+            test.leftOrNull()
+        }.toNonEmptyListOrNull()
```

Comment:
very hacky piece of code

Reply:
I am not sure what i should change since the IorNel return ask for a NonEmptyList.